src/ert/_c_wrappers/analysis/analysis_module.py

```diff
@@ -1,162 +1,230 @@
-            return self.getDouble(name)
-        elif variable_type == bool:
-            return self.getBool(name)
-        elif variable_type == int:
-            return self.getInt(name)
+    if mode == AnalysisMode.ENSEMBLE_SMOOTHER:
+        return es_variables
+    return ies_variables
+
+
+class AnalysisModule:
+    DEPRECATED_KEYS = ["USE_EE", "USE_GE"]
+    TRUNC_ALTERNATE_KEYS = ["ENKF_NCOMP", "ENKF_SUBSPACE_DIMENSION"]
+    SPECIAL_KEYS = ["INVERSION", *TRUNC_ALTERNATE_KEYS, *DEPRECATED_KEYS]
+
+    def __init__(
+        self,
+        mode: AnalysisMode,
+        name: str,
+        variables: Dict[str, "VariableInfo"],
+        iterable: bool,
+    ):
+        self.mode = mode
+        self.name = name
+        self.iterable = iterable
+        self._variables = variables
+
+    @classmethod
+    def ens_smoother_module(cls, name: AnalysisMode = AnalysisMode.ENSEMBLE_SMOOTHER):
+        return cls(
+            mode=AnalysisMode.ENSEMBLE_SMOOTHER,
+            name=name,
+            variables=get_mode_variables(AnalysisMode.ENSEMBLE_SMOOTHER),
+            iterable=False,
+        )
+
+    @classmethod
+    def iterated_ens_smoother_module(
+        cls, name: AnalysisMode = AnalysisMode.ITERATED_ENSEMBLE_SMOOTHER
+    ):
+        return cls(
+            mode=AnalysisMode.ITERATED_ENSEMBLE_SMOOTHER,
+            name=name,
+            variables=get_mode_variables(AnalysisMode.ITERATED_ENSEMBLE_SMOOTHER),
+            iterable=True,
+        )
+
+    def get_variable_names(self) -> List[str]:
+        return list(self._variables.keys())
+
+    def get_variable_value(self, name) -> Union[int, float, bool]:
+        if name in self._variables:
+            return self._variables[name]["value"]
+        raise KeyError(f"Variable {name} not found in module {self.name}")
+
+    def variable_value_dict(self) -> Dict[str, Union[float, int]]:
+        return {name: var["value"] for name, var in self._variables.items()}
+
+    def handle_special_key_set(self, var_name, value):
+        if var_name in self.DEPRECATED_KEYS:
+            logger.warning(
+                f"{var_name} settings have been removed"
+                f"use the INVERSION setting instead"
+            )
+        elif var_name == "INVERSION":
+            inversion_str_map = {
+                "EXACT": 0,
+                "SUBSPACE_EXACT_R": 1,
+                "SUBSPACE_EE_R": 2,
+                "SUBSPACE_RE": 3,
+            }
+            if value in inversion_str_map:
+                self.set_var("IES_INVERSION", inversion_str_map[value])
+            else:
+                logger.warning(
+                    f"Unknown value {value} used for INVERSION key"
+                    f"supported values are {list(inversion_str_map.keys())}"
+                )
+        elif var_name in self.TRUNC_ALTERNATE_KEYS:
+            self.set_var("ENKF_TRUNCATION", value)
+
+    def set_var(self, var_name: str, value: Union[float, int, bool, str]):
```

Comment:
It is possible to type this more succinctly with the overload

```python
from typing import overload

int_var_names = [v["name"] for v in ies_variables.values() if v["type"] == int]
float_var_names = [v["name"] for v in ies_variables.values() if v["type"] == float]

@overload
def set_var(self, var_name: Literal[int_var_names], value: int): ...
@overload
def set_var(self, var_name: Literal[float_var_names], value: float): ...
def set_var(self, var_name, value):
    if var_name in self.DEPRECATED_KEYS:
        logger.warning(...
```

Reply:
I am sorry my I do not understand how the suggestion is more succinct. 

`@overload` has never been used in the ERT codebase before, I don't really want to be the first to start adding it to the codebase.

I never used overloaded functions in python before so I find it a bit hard to understand what you are suggesting.

Reply:
We can leave that typing for now, and I can propose a change later